pkg/agent/plugin/nodeattestor/tpmdevid/tpmutil/signingkey.go

```diff
@@ -0,0 +1,180 @@
+package tpmutil
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/google/go-tpm-tools/tpm2tools"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+	"github.com/hashicorp/go-hclog"
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
+)
+
+// maxAttempts indicates the max number retries for running TPM commands when
+// TPM responds with a tpm2.RCRetry code.
+const maxAttempts = 10
+
+// SigningKey represents a TPM loaded key
+type SigningKey struct {
+	Handle     tpmutil.Handle
+	sigHashAlg tpm2.Algorithm
+	rw         io.ReadWriter
+	log        hclog.Logger
+}
+
+// LoadSigningKey loads the given keypair into the provided TPM
+func LoadSigningKey(rw io.ReadWriter, pubKey, privKey []byte, log hclog.Logger) (*SigningKey, error) {
+	pub, err := tpm2.DecodePublic(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Public decoding failed: %w", err)
+	}
+
+	canSign := pub.Attributes&tpm2.FlagSign != 0
+	if !canSign {
+		return nil, fmt.Errorf("not a signing key")
+	}
+
+	var sigHashAlg tpm2.Algorithm
+	var srkTemplate tpm2.Public
+	switch pub.Type {
+	case tpm2.AlgRSA:
+		srkTemplate = SRKTemplateHighRSA()
+		rsaParams := pub.RSAParameters
+		if rsaParams != nil {
+			sigHashAlg = rsaParams.Sign.Hash
+		}
+
+	case tpm2.AlgECC:
+		srkTemplate = tpm2tools.SRKTemplateECC()
+		eccParams := pub.ECCParameters
+		if eccParams != nil {
+			sigHashAlg = eccParams.Sign.Hash
+		}
+
+	default:
+		return nil, fmt.Errorf("bad key type: 0x%04x", pub.Type)
+	}
+
+	if sigHashAlg.IsNull() {
+		return nil, fmt.Errorf("signature hash algorithm is NULL")
+	}
+
+	srk, err := tpm2tools.NewKey(rw, tpm2.HandleOwner, srkTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("SRK creation failed: %w", err)
+	}
+
+	defer srk.Close()
+
+	keyHandle, _, err := tpm2.Load(rw, srk.Handle(), "", pubKey, privKey)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Load failed: %w", err)
+	}
+
+	return &SigningKey{
+		Handle:     keyHandle,
+		sigHashAlg: sigHashAlg,
+		rw:         rw,
+		log:        log,
+	}, nil
+}
+
+// Close removes the key from the TPM
+func (k *SigningKey) Close() error {
+	return tpm2.FlushContext(k.rw, k.Handle)
+}
+
+// Sign request the TPM to sign the given data using this key
+func (k *SigningKey) Sign(data []byte) ([]byte, error) {
+	digest, token, err := tpm2.Hash(k.rw, k.sigHashAlg, data, tpm2.HandlePlatform)
+	if err != nil {
+		return nil, fmt.Errorf("tpm2.Hash failed: %w", err)
+	}
+
+	var sig *tpm2.Signature
+	for i := 0; i <= maxAttempts; i++ {
+		sig, err = tpm2.Sign(k.rw, k.Handle, "", digest, token, nil)
+		switch {
+		case err == nil:
+			break
+
+		case isRetry(err):
+			if i == maxAttempts {
+				return nil, fmt.Errorf("max attempts reached: %w", err)
```

Comment:
It would be useful to include information about the action involved, otherwise the error may be difficult to understand. Maybe something like `max attempts reached while trying to sign the given data`.